src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
 	return 1;
 }
 
+// Add a reference from an IGCUserPeer jobject to another jobject
 static mono_bool
-add_reference (JNIEnv *env, MonoObject *obj, MonoJavaGCBridgeInfo *bridge_info, MonoObject *reffed_obj)
+add_reference_jobject (JNIEnv *env, jobject handle, jobject reffed_handle)
 {
 	jclass java_class;
 	jmethodID add_method_id;
-	void *handle, *reffed_handle;
-#if DEBUG
-	MonoClass *klass;
-	klass = mono.mono_object_get_class (obj);
-#endif
 
-	mono.mono_field_get_value (obj, bridge_info->handle, &handle);
 	java_class = (*env)->GetObjectClass (env, handle);
 	add_method_id = (*env)->GetMethodID (env, java_class, "monodroidAddReference", "(Ljava/lang/Object;)V");
 	if (add_method_id) {
-		mono.mono_field_get_value (reffed_obj, bridge_info->handle, &reffed_handle);
 		(*env)->CallVoidMethod (env, handle, add_method_id, reffed_handle);
 		(*env)->DeleteLocalRef (env, java_class);
-#if DEBUG
-		if (gc_spew_enabled)
-			log_warn (LOG_GC, "added reference for object of class %s.%s to object of class %s.%s",
-					mono.mono_class_get_namespace (klass),
-					mono.mono_class_get_name (klass),
-					mono.mono_class_get_namespace (mono.mono_object_get_class (reffed_obj)),
-					mono.mono_class_get_name (mono.mono_object_get_class (reffed_obj)));
-#endif
+
 		return 1;
 	}
 
 	(*env)->ExceptionClear (env);
-#if DEBUG
-	if (gc_spew_enabled)
-		log_error (LOG_GC, "Missing monodroidAddReference method for object of class %s.%s",
-				mono.mono_class_get_namespace (klass),
-				mono.mono_class_get_name (klass));
-#endif
 	(*env)->DeleteLocalRef (env, java_class);
 	return 0;
 }
 
+// add_reference can work with objects which are either MonoObjects with java peers, or raw jobjects
+typedef struct {
+	mono_bool is_mono_object;
+	union {
+		MonoObject *obj;
+		jobject jobj;
+	};
+} AddReferenceTarget;
+
+// These will be loaded as needed and persist between GCs
+// FIXME: This code assumes it is totally safe to hold onto these GREFs forever. Can mono.android.jar ever be unloaded?
+static jobject   jni_arraylist, jni_gcuserpeer;
```

Comment:
Convention is:

1. Prefix based on the Java class name (excluding package), preserving case. **Good**: `ArrayList`. **Bad**: `arraylist`.

2. Underscore.

3. "Member name": the field/method name, or `ctor` for constructors, or `class` for `jclass` values.

These are also usually done one per line, organized by type:

Thus:

    static jclass     ArrayList_class;
    static jmethodID  ArrayList_ctor;
    static jmethodID  ArrayList_get;
    static jmethodID  ArrayList_add;

    static jclass     GCUserPeer_class;
    static jmethodID  GCUserPeer_ctor;